staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go

```diff
@@ -103,20 +103,23 @@ func BeforeDelete(strategy RESTDeleteStrategy, ctx context.Context, obj runtime.
 		// 2. Delete the object from storage.
 		// If the update succeeds, but the delete fails (network error, internal storage error, etc.),
 		// a resource was previously left in a state that was non-recoverable.  We
-		// check if the existing stored resource has a grace period as 0 and if so
+		// check if the existing stored resource has a grace period <= 0 and if so
 		// attempt to delete immediately in order to recover from this scenario.
-		if objectMeta.GetDeletionGracePeriodSeconds() == nil || *objectMeta.GetDeletionGracePeriodSeconds() == 0 {
+		if objectMeta.GetDeletionGracePeriodSeconds() == nil || *objectMeta.GetDeletionGracePeriodSeconds() <= 0 {
```

Comment:
We don't want this change as we'll go from graceful to forceful shutdown with a negative number.

Reply:
A negative grace period passed to the delete API is not valid. 0 is the closest valid value, so is what I would expect.

Reply:
A negative grace period passed to the delete API is not valid. 0 is the closest valid value, so is what I would expect.

Reply:
I disagree.  If you were looking at the drain command as a reference rather than the API and missed the bit where only non-negative integers are passed, you might assume that sending a negative is valid: https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/drain/drain.go#L194

To me, a negative value is clearly intended to use the pod's default value.

Reply:
FWIW in the kubelet I don't think it's possible to delete with a grace period below the minimum default, this gets validated and overridden here:

https://github.com/kubernetes/kubernetes/blob/1fb27c4b5d6c9d8f508c7b4bbcf96244d6b65449/pkg/kubelet/pod_workers.go#L302-L309

https://github.com/kubernetes/kubernetes/blob/1fb27c4b5d6c9d8f508c7b4bbcf96244d6b65449/pkg/kubelet/kuberuntime/kuberuntime_container.go#L635-L638